data-processing/common/expand_macros.py

```diff
@@ -0,0 +1,492 @@
+    + br"\(If this.*?, it appeared in (?P<path>.*?) "
+    + br"from line (?P<start_line>\d+), col (?P<start_column>\d+) "
+    + br"to line (?P<end_line>\d+), col (?P<end_column>\d+)\)\.",
+    flags=re.MULTILINE | re.DOTALL,
+)
+
+TOKEN_PATTERN_PREFIX = b"Expansion token:"
+TOKEN_PATTERN = re.compile(
+    b"Expansion token: (?P<token>.*?) "
+    + br"\(object ID (?P<object_id>.*?)\)\. "
+    + br"Category: (?P<category_code>\d+)\. "
+    + br"Expandable: (?:true|false)\.",
+    flags=re.MULTILINE | re.DOTALL,
+)
+
+ARGUMENT_PATTERN_PREFIX = b"Argument token (from file):"
+ARGUMENT_PATTERN = re.compile(
+    br"Argument token \(from file\): "
+    + br'".*?" \(object ID: (?P<object_id>.*?)\)\. '
+    + br"\(source file (?P<path>.*?), "
+    + br"from line (?P<start_line>\d+) col (?P<start_column>\d+) "
+    + br"to line (?P<end_line>\d+) col (?P<end_column>\d+)\)\.",
+    flags=re.MULTILINE | re.DOTALL,
+)
+
+END_PATTERN_PREFIX = b"End of expansion "
+END_PATTERN = re.compile(
+    b"End of expansion "
+    + br"\(object ID: (?P<object_id>.*?)\)\. "
+    + br"Current expansion depth: \d+\. "
+    + br"Expansion: .+?\.",
+    flags=re.MULTILINE | re.DOTALL,
+)
+
+
+@dataclass
+class LogPattern:
+    prefix: bytes
+    capture: Pattern[bytes]
+
+
+patterns: Dict[str, LogPattern] = {
+    "definition": LogPattern(DEFINITION_PATTERN_PREFIX, DEFINITION_PATTERN),
+    "start-expansion": LogPattern(START_PATTERN_PREFIX, START_PATTERN),
+    "add-expansion-token": LogPattern(TOKEN_PATTERN_PREFIX, TOKEN_PATTERN),
+    "read-argument": LogPattern(ARGUMENT_PATTERN_PREFIX, ARGUMENT_PATTERN),
+    "end-expansion": LogPattern(END_PATTERN_PREFIX, END_PATTERN),
+}
+
+
+@dataclass
+class Expansion:
+    """
+    The output of the method for detecting expansions. Includes all the information that should
+    be needed to replace each appearance of a macro with its expansion---the position of the
+    macro and its arguments, and the text (as bytes) to replace it with.
+    """
+
+    macro_name: bytes
+    " The token containing the macro name (without arguments). Should begin with '\\'. "
+
+    start_line: int
+    start_col: int
+    end_line: int
+    end_col: int
+    """
+    'start' and 'end' represent the range of all characters that should be replaced with an
+    expansion. This should include both the control sequence and its arguments:
+    * For a simple macro without arguments, this will be just the control sequence itself
+    (e.g., just '\\X' for macro '\\X' defined as '\\def\\X{X}').
+    * For a macro with arguments, this range will cover both the control sequence and its
+    arguments (e.g., all of '\\c{X}' for the macro '\\c' defined as '\\def\\c#1{\\mathcal{C}}')).
+
+    For 'start_line' and 'end_line', the first line in a file is 1 (not 0). For 'start_col' and
+    'end_col', the first character on the line is at col 0 (not 1, in contrast to the line number).
+    """
+
+    expansion: bytes
+
+
+def detect_expansions(
```

Comment:
totally optional, but this is quite a beefy function, I wonder if it would be possible to break it up into smaller chunks so the overall shape of what it's doing could be seen at a glance and then you could dive into the broken out functions to see details of how a particular type is processed for instance?

Reply:
I was thinking about this! I will take a stab at breaking this into a few functions, probably as part of a class (so that state can be shared). If I can do it in a time-boxed interval, we can review it as an alternative.

Reply:
@aschokking hopefully the changes below (from monolithic method to class with half a dozen methods) makes the code a bit easier to read / navigate by chunking it up. Want to take a look?

Reply:
Thanks Andrew, `detect_expansions ` is definitely a lot easier to read now!